project-templates/java-emf-theia/glsp-server/src/main/java/org/eclipse/glsp/example/javaemf/emf/EMFGModelState.java

```diff
@@ -0,0 +1,175 @@
+import java.util.Optional;
+
+import com.google.inject.Inject;
+
+import org.apache.log4j.LogManager;
+import org.apache.log4j.Logger;
+import org.eclipse.emf.common.command.BasicCommandStack;
+import org.eclipse.emf.common.command.CommandStack;
+import org.eclipse.emf.common.util.EList;
+import org.eclipse.emf.common.util.URI;
+import org.eclipse.emf.ecore.EObject;
+import org.eclipse.emf.ecore.resource.Resource;
+import org.eclipse.emf.ecore.resource.ResourceSet;
+import org.eclipse.emf.ecore.util.EcoreUtil;
+import org.eclipse.emf.edit.domain.AdapterFactoryEditingDomain;
+import org.eclipse.emfcloud.modelserver.glsp.notation.Diagram;
+import org.eclipse.glsp.graph.GModelRoot;
+import org.eclipse.glsp.server.model.DefaultGModelState;
+import org.eclipse.glsp.server.model.GModelState;
+import org.eclipse.glsp.server.session.ClientSession;
+import org.eclipse.glsp.server.session.ClientSessionListener;
+import org.eclipse.glsp.server.session.ClientSessionManager;
+import org.eclipse.glsp.server.types.GLSPServerException;
+import org.eclipse.glsp.server.utils.ClientOptionsUtil;
+
+/**
+ * This state represents the status of the diagram based on the {@link GModelRoot}, contains the {@link EMFModelIndex},
+ * is able to execute commands via its {@link CommandStack} and is registered as {@link ClientSessionListener}
+ * to be able to reset the EMF resources on diagram close.
+ */
+public abstract class EMFGModelState extends DefaultGModelState implements ClientSessionListener {
+
+    private static Logger LOGGER = LogManager.getLogger(EMFGModelState.class.getSimpleName());
+
+    @Inject
+    protected ClientSessionManager clientSessionManager;
+
+    protected ResourceSet resourceSet;
+    protected AdapterFactoryEditingDomain editingDomain;
+
+    public abstract EObject getSemanticModel();
+
+    public abstract Diagram getNotationModel();
+
+    @Inject
+    public void init() {
+        this.clientSessionManager.addListener(this, this.clientId);
+        initializeCommandStack();
+        setupResourceSet();
+    }
+
+    @Override
+    public void setRoot(final GModelRoot newRoot) { this.currentModel = newRoot; }
+
+    @Override
+    protected void initializeCommandStack() {
+        if (commandStack != null) {
+            commandStack.flush();
+        }
+        commandStack = new BasicCommandStack();
+    }
+
+    public ResourceSet getResourceSet() { return resourceSet; }
+
+    public AdapterFactoryEditingDomain getEditingDomain() { return editingDomain; }
+
+    protected abstract void setupResourceSet();
+
+    public abstract void loadResources();
+
+    @Override
+    public EMFModelIndex getIndex() { return EMFModelIndex.get(getRoot()); }
+
+    public static EMFGModelState getModelState(final GModelState state) {
+        if (!(state instanceof EMFGModelState)) {
+            throw new IllegalArgumentException("Argument must be a EMFGModelState");
+        }
+        return ((EMFGModelState) state);
+    }
```

Comment:
I'm not sure whether this is ever needed. I think every other class can get the concrete subtype injected.